ESM build with Deno support

alexjeffburke opened the pull request:
This PR worked on by @papandreou and myself implements a working port to Unexpected to deno. The changes involved are the creation of an esm build and some porting work to enable colour terminal support in `magicpen` (https://github.com/sunesimonsen/magicpen/pull/26) plus `unexpected-magicpen` (https://github.com/unexpectedjs/unexpected-magicpen/pull/19) to support the test suite.

As for the why of it - as a very new platform deno doesn’t yet have a fully featured assertion library and our discussions noted that, should it be possible to do so without being too invasive, we might gain some interesting exposure on that platform. The challenge was to see what the porting effort would be - and this showcases it.

alexjeffburke commented on .gitignore:
```diff
@@ -8,3 +8,5 @@
 /.nyc_output/
 /unexpected.js
 /unexpected.js.map
+/unexpected-deno.js
```
If we do end up generating this, I wonder if we should just call the file `esm` and the. allow users to `import expect from “unexpected/esm"`.

Also need to check what the options are for making these files accessible to the deno registry.

papandreou replied:
`import expect from “unexpected/esm"` won't really help deno users, as they aren't using npm. It seems like we'll need to get unexpected added to https://github.com/denoland/registry/blob/master/src/database.json, check the generated file into git, and then hand out an url like https://deno.land/x/unexpected/unexpected-deno.js

alexjeffburke replied:
That’s a fair point - currently the deno registry is limited to pointing you to the project but the you must be able to import a file that is checked in. That said, the esm build is still generic so if we do check it in I think we should consider renaming :)

papandreou replied:
Absolutely, not a lot of thought went into the name :)

alexjeffburke commented on test-deno.js:
```diff
@@ -0,0 +1,3 @@
+import expect from './unexpected-deno.js';
+
```
Ideally this would execute the full battery of tests we have - got some early work in the direction of trying to get the real suite running.

alexjeffburke commented:
@papandreou I think this is ready for a real review. I've spoken to @sunesimonsen who mentioned he would be ok moving forward with this if it passes your review. If the changes here are ok, then I'll move on to ushering the requisite changes into `magicpen`. The only other requirement is an ESM build of unexpected-magicpen, but aisde from a switch to rollup I don't think that is controversial.

papandreou commented on test/types/function-type.spec.js:
```diff
@@ -44,120 +44,123 @@ describe('function type', () => {
+      /* eslint-disable no-unused-vars */
+      expect(
+        // prettier-ignore
+        function() { var a = 123;a = 456; },
```
Seems like the reason why these tests don't pass in deno is that the transpilation of the tests realizes that the code inside the function can be optimized away. Could you try changing it to something with side effects, such as the below?

```suggestion
        function() { var a = 123;alert(a); },
```

... or find a way to make the transpilation leave it alone?

alexjeffburke replied:
Hmm interesting - the symptoms I saw was when this was inspected it came out as “native code” so I’d assumed it was a bug in Deno. I’d much rather these ran too so will see why I can do!

papandreou commented on Makefile:
```diff
@@ -1,8 +1,8 @@
-.PHONY: unexpected.js
-.SECONDARY: unexpected.js.map
+TARGETS ?= unexpected.js unexpected.js.map unexpected.esm.js unexpected.esm.js.map
+.PHONY: unexpected.js unexpected-deno.js
```
```suggestion
.PHONY: unexpected.js unexpected.esm.js
```

papandreou commented on Makefile:
```diff
@@ -1,8 +1,8 @@
-.SECONDARY: unexpected.js.map
+TARGETS ?= unexpected.js unexpected.js.map unexpected.esm.js unexpected.esm.js.map
+.PHONY: unexpected.js unexpected-deno.js
+.SECONDARY: unexpected.js.map unexpected-deno.js.map
```
```suggestion
.SECONDARY: unexpected.js.map unexpected.esm.js.map
```

papandreou commented on .prettierignore:
```diff
@@ -3,6 +3,7 @@
 /coverage/
 /site-build/
 /unexpected.js
+/unexpected-deno.js
```
```suggestion
/unexpected.esm.js
```

papandreou commented on .eslintignore:
```diff
@@ -4,3 +4,4 @@
 /coverage/
 /site-build/
 /unexpected.js
+/unexpected-deno.js
```
```suggestion
/unexpected.esm.js
```

alexjeffburke commented:
@papandrepu thanks very much for the review! Will address those bits but then I think we’re in pretty decent shape :) And just to say I’m also trying to ensure we don’t have to return to checking in assets: https://github.com/denoland/registry/pull/152

alexjeffburke commented:
I've addressed all the comments including removing the conditional block around the function inspection tests which now all pass. The latest build passed using the released versions `magicpen` and `unexpected-magicpen` 😃

papandreou commented on test/types/function-type.spec.js:
```diff
@@ -84,79 +85,72 @@ describe('function type', () => {
   /* eslint-disable no-unused-vars */
   function singleLineWithComment() {
```
I guess the function name should change to `twoLineWithComment` :)

papandreou commented on package.json:
```diff
@@ -63,18 +63,21 @@
     "serve": "*",
     "unexpected-documentation-site-generator": "^6.0.0",
-    "unexpected-magicpen": "^2.0.0",
+    "unexpected-magicpen": "2.1.0",
```
```suggestion
    "unexpected-magicpen": "^2.1.0",
```

papandreou commented on package.json:
```diff
@@ -28,7 +28,7 @@
     "greedy-interval-packer": "1.2.0",
     "leven": "2.1.0",
-    "magicpen": "^6.0.2",
+    "magicpen": "6.1.0",
```
```suggestion
    "magicpen": "^6.1.0",
```
